pkg/node/register.go

```diff
@@ -0,0 +1,160 @@
+// Copyright 2018 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package node
+
+import (
+	"encoding/json"
+	"fmt"
+	"path"
+	"sync"
+	"time"
+
+	"github.com/cilium/cilium/pkg/controller"
+	"github.com/cilium/cilium/pkg/kvstore"
+
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	// localNodeRegistrationInterval is the interval in which the local
+	// node is re-registered with the kvstore and BPF maps
+	localNodeRegistrationInterval = time.Duration(2) * time.Minute
+
+	// timeout to wait for initial list of nodes
+	listTimeout = time.Duration(30) * time.Second
+)
+
+var (
+	// nodesPath is the path to where nodes are stored in the kvstore
+	nodesPath = path.Join(kvstore.BaseKeyPrefix, "state", "nodes", "v1")
+
+	// controllers contains all controllers required for node management
+	controllers controller.Manager
+)
+
+var registerOnce sync.Once
+
+// RegisterLocalNode registers the local node in the cluster
+func RegisterLocalNode() error {
+	n := GetLocalNode()
+	n.getLogger().Debug("Registering local node")
+
+	// Perform an initial blocking node registration to ensure that the
+	// agent fails when we can't register
+	err := n.registerLocalNode()
+	if err != nil {
+		return fmt.Errorf("unable to register local node: %s", err)
+	}
+
+	registerOnce.Do(func() {
+		// Following the initial blocking registration, schedule a controller
+		// to follow-up and ensure the kvstore entry gets re-added
+		go func() {
+			time.Sleep(localNodeRegistrationInterval)
```

Comment:
the `sync.Once` blocks all calls until the first call finishes. You could call the first `registerLocalNode` inside this `sync.Do`